network/src/protocols/network/mod.rs

```diff
@@ -172,22 +174,58 @@ pub trait NewNetworkEvents {
     fn new(
         peer_mgr_notifs_rx: aptos_channel::Receiver<(PeerId, ProtocolId), PeerManagerNotification>,
         connection_notifs_rx: aptos_channel::Receiver<PeerId, ConnectionNotification>,
+        bounded_executor: BoundedExecutor,
     ) -> Self;
 }
 
-impl<TMessage: Message> NewNetworkEvents for NetworkEvents<TMessage> {
+impl<TMessage: Message + Send + 'static> NewNetworkEvents for NetworkEvents<TMessage> {
     fn new(
-        peer_mgr_notifs_rx: aptos_channel::Receiver<(PeerId, ProtocolId), PeerManagerNotification>,
+        mut peer_mgr_notifs_rx: aptos_channel::Receiver<
+            (PeerId, ProtocolId),
+            PeerManagerNotification,
+        >,
         connection_notifs_rx: aptos_channel::Receiver<PeerId, ConnectionNotification>,
+        bounded_executor: BoundedExecutor,
     ) -> Self {
-        let data_event_stream = peer_mgr_notifs_rx.filter_map(
-            peer_mgr_notif_to_event
-                as fn(PeerManagerNotification) -> future::Ready<Option<Event<TMessage>>>,
-        );
+        // Create a channel for deserialized messages
+        let (deserialized_message_sender, deserialized_message_receiver) =
+            mpsc::channel(MAX_DESERIALIZATION_QUEUE_SIZE);
+
+        // Deserialize the peer manager notifications in parallel (for each
+        // network application) and send them to the receiver. Note: this
+        // may cause out of order message delivery.
+        Handle::current().spawn(async move {
+            // For each notification, spawn a new task to deserialize the message
+            while let Some(peer_manager_notification) = peer_mgr_notifs_rx.next().await {
+                let mut deserialized_message_sender = deserialized_message_sender.clone();
+                bounded_executor
+                    .spawn(async move {
+                        if let Some(deserialized_message) =
+                            peer_mgr_notif_to_event(peer_manager_notification).await
```

Comment:
`peer_mgr_notif_to_event()` no longer needs to return a future (it's actually a blocking function)

And using `spawn()` in hope these heavy operations will happen in different threads seems not the rightest thing to do 😂

Reply:
> peer_mgr_notif_to_event() no longer needs to return a future (it's actually a blocking function)

Aah, nice -- missed this! Thanks :D Also, I changed to use `spawn_blocking` to satisfy the heavy operation requirements.